### DISCUSSION TITLE
How do nodes subscribe for changes

### DISCUSSION THREAD

**Discussion by computinglife** [rqlite/rqlite#1813] [General]
Hi,  

While reading the doc on Read-Only nodes, i came across this line "read-only nodes still subscribe to the stream of committed log entries broadcast by the Leader" 

Can you please point me to the architecture here? How the subscription works and what the tech is ? 

Thanks in advance 
Arun

**Comment by otoolep** [comment]
The process is the same as the code used to stream updates to the voting (normal) nodes. The Leader keeps track of what each node has received and sends Raft logs entries as normal. The only difference is read-only nodes don't vote and just accept changes unconditionally from the Leader.

**Comment by computinglife** [comment]
Thank you - I was trying to understand what is the underlying communication stack 

From reading the docs i think the requests to the /execute end point are being streamed using ProtoBuf API that you have described in the docs?

**Comment by otoolep** [comment]
It uses a custom protocol over a persistent TCP connection. For more details you would need to study the Raft source code:
GitHub - hashicorp/raft: Golang implementation of the Raft consensus protocol 
  
|  
|   
|   
|   |    |

   |

  |
|  
|    |  
GitHub - hashicorp&#47;raft: Golang implementation of the Raft consensus protocol
 
Golang implementation of the Raft consensus protocol - hashicorp&#47;raft
  |   |

  |

  |

  



Sent from Yahoo Mail for iPhone


On Wednesday, June 26, 2024, 9:53 AM, computinglife ***@***.***> wrote:




Thank you - I was trying to understand what is the underlying communication stack ? Are they are communicating the commits to each other as an HTTP request ? Or is there any other custom protocol in use here with which they talk to each other?

**Comment by computinglife** [comment]
Thank you - that clears it up.
